<div class="alert alert-success"><div class='alert btn-success'>
Павел, 😊буду рада, если смогла быть тебе полезной!)<br>
Критичных ошибок нет. Молодец!)<br>
С удовольствием принимаю проект!<br> 

</div></div>

Очень рекомендую посмотреть видео от Глеба Михайлова!)<br>
**[Линейная Регрессия для Дата Саентиста](https://www.youtube.com/watch?v=QZJ94igWVxQ&t=19s)**     <br> 
    
**[Вот здесь есть отличное объяснение с примером по мультиколлинеарности.](https://habr.com/ru/company/akbarsdigital/blog/592493/)** <br>
    
**[И здесь интересная статься - коротко о мультиколлинеарности и что с этим делать.](https://medium.com/analytics-vidhya/removing-multi-collinearity-for-linear-and-logistic-regression-f1fa744f3666)** <br>   
   
**[Если интересна статистика, можно посмотреть короткие видео-лекции Карпова на степике](https://stepik.org/lesson/8095/step/1?unit=1371)**    <br>


<br>
<div class="alert alert-success"><div class='alert btn-success'>    
Надеюсь, ты найдешь для себя что-то интересное и полезное😊  <br>
Желаю тебе удачи в дальнейшей учебе! 😉 </div></div>

Павел, привет! Меня зовут Ирина Литвинова. Сегодня я проверю твой проект.<br>
<br> Ты можешь обращаться ко мне на "ты"😏 Надеюсь, ты не будешь против, если и я буду обращаться к тебе на "ты". Но если для тебя это неприемлeмо, дай знать, пожалуйста) <br>
<br> Мне очень важна твоя реакция на мои комментарии, особенно <font color='red'>красные</font>. Можешь реагировать, например, так ('исправлено', ...)<br>
<br> Ты можешь реагировать на мои комментарии своими, выделяя их цветами и наиболее понравившимся тебе способом оформления, но явно  отличающимся от моих комментариев. <br>
Это нужно, чтобы не создавалась путаница😉<br>
Например, так:<br>
    
<div class="alert alert-block alert-info"><b>Комментарий студента</b>
    
<b>Изменения:</b> Были внесены следующие изменения ...<br></div> 


<br> Пожалуйста, не удаляй мои комментарии, так как они повышают качество повторного ревью и делают его быстрее)<br>
<br> Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой руководитель будет поступать также, и я пытаюсь подготовить тебя именно к работе. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку!🤓<br>

Для твоего удобства мои комментарии будут в <font color='green'>зеленой</font>, <font color='orange'>желтой</font> или <font color='red'>красной</font> рамках:<br>
<br>

<div class="alert alert-block alert-danger"><b>Комментарий ревьюера</b>
    
<b>Ошибка:</b><br>
Если требуются исправления. Работа не может быть принята с красными комментариями.<br></div>

<div class="alert alert-block alert-warning"><b>Комментарий ревьюера</b>
    
<b>Совет:</b><br>
Если можно немного улучшить.<br></div>

<div class="alert alert-block alert-success"><b>Комментарий ревьюера</b>
    
<b>Успех:</b><br>
Если все сделано отлично.<br></div>  

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

<div class="alert alert-block alert-success"><b>Комментарий ревьюера</b>
    
<b>Успех:</b><br>
👍 Есть введение в проект. Отлично!<br>

</div> 

### Загружаем библиотеки:

In [50]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
from sklearn.model_selection import train_test_split 
from sklearn.linear_model import LogisticRegression
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler

<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b>
    
<b>Успех:</b><br>
👍 Здорово, что импортируешь необходимые библиотеки в первой ячейке. Так сразу понятно какие используются библиотеки и можно легко настроить свое окружение при необходимости. Достаточно оставлять только те, которые используются в проекте ))
</div> 

## Загрузка и подготовка данных

Загрузка данных:

In [51]:
try:
    region1 = pd.read_csv('geo_data_0.csv')
    region2 = pd.read_csv('geo_data_1.csv')
    region3 = pd.read_csv('geo_data_2.csv')
    print('датасет загружен локально')
except:
    region1 = pd.read_csv('/datasets/geo_data_0.csv')
    region2 = pd.read_csv('/datasets/geo_data_1.csv')
    region3 = pd.read_csv('/datasets/geo_data_2.csv')
    print('датасет загружен с сервера')

датасет загружен локально


<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b><br>
    
 Хорошо что используешь `try/except`, чтобы избежать ошибок загрузки. `try/except` достаточно для загрузки файлов, нет необходимости в лишнем коде 👇
```python
region1 = pd.read_csv('/datasets/geo_data_0.csv')
region2 = pd.read_csv('/datasets/geo_data_1.csv')
region3 = pd.read_csv('/datasets/geo_data_2.csv') 
```    
    
</div> 

Информация о датасетах:

In [52]:
display(region1.describe(), region2.describe(), region3.describe())


f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

Аномалий нет, проверим на дубликаты:

In [53]:
print(region1.id.duplicated().sum(),region2.id.duplicated().sum(),region3.id.duplicated().sum())

10 4 4


Есть дубликаты в id - в этом случае это скорее всего ошибка, поэтому нужно избавиться от дубликатов. Их совсем не много, ничем не рискуем.

<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b>
    
<b>Успех:</b><br>
👍 Хорошо, что проверяешь дубликаты по `id`. Так как нам не ясна природа таких дубликатов, то их можно как удалить, так и оставить. Но о таких фактах всегда надо сообщать коллегам.
</div> 

In [54]:
region1 = region1.drop_duplicates(subset='id').reset_index(drop=True)
region2 = region2.drop_duplicates(subset='id').reset_index(drop=True)
region3 = region3.drop_duplicates(subset='id').reset_index(drop=True)

In [55]:
# Проверим
print(region1.id.duplicated().sum(),region2.id.duplicated().sum(),region3.id.duplicated().sum())

0 0 0


In [56]:
#Посмотрим еще раз на датасеты:
display(region1.info(), region2.info(), region3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99990 entries, 0 to 99989
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       99990 non-null  object 
 1   f0       99990 non-null  float64
 2   f1       99990 non-null  float64
 3   f2       99990 non-null  float64
 4   product  99990 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99996 entries, 0 to 99995
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       99996 non-null  object 
 1   f0       99996 non-null  float64
 2   f1       99996 non-null  float64
 3   f2       99996 non-null  float64
 4   product  99996 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99996 entries, 0 to 99995
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
--

None

None

None

<div class="alert alert-block alert-warning"><b>Комментарий ревьюера</b>
    
<b>Совет:</b><br>
Не забывай, пожалуйста, в конце каждого раздела формулировать выводы. Код - инструмент, выводы - цель.
</div>

<div class="alert alert-block alert-warning"><b>Комментарий ревьюера</b>
    
<b>Совет:</b><br>
Павел,     
* Чтобы лучше узнать данные, можно посмотреть распределение признаков и целевой переменной для каждого региона, т.е. построить гистограмму.    
* Еще важно уделять внимание мультиколлинеарности признаков в линейных моделях.<br>
Мультиколлинеарность — явление, при котором наблюдается сильная корреляция между признаками, что может негативно отразиться на модели.<br>

Рекомендую посмотреть отрывок из курса Статистики - ["Корреляция и регрессия".](https://stepik.org/lesson/8086/step/1?unit=1365) <br>
Для удобной визуализации можно использовать - ["Тепловая карта корреляции".](https://www.delftstack.com/ru/howto/seaborn/correlation-heatplot-seaborn-python/) <br>
</div>

## Обучение и проверка модели

In [57]:
# Создадим выборки для каждого региона:
features_r1 = region1.drop(['id', 'product'], axis=1)
target_r1 = region1['product'].reset_index(drop=True)

features_r2 = region2.drop(['id','product'], axis=1)
target_r2 = region2['product'].reset_index(drop=True)

features_r3 = region3.drop(['id', 'product'], axis=1)
target_r3 = region3['product'].reset_index(drop=True)

<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b>
    
<b>Успех:</b><br>
👍 `id`  - уникальный идентификатор скважины, он не является признаком, поэтому его действительно лучше удалить. 
</div> 

In [58]:
# Разделим выборки
features_train_r1, features_valid_r1, target_train_r1, target_valid_r1 = \
train_test_split(features_r1, target_r1, test_size=0.25, random_state=12345)

features_train_r2, features_valid_r2, target_train_r2, target_valid_r2 = \
train_test_split(features_r2, target_r2, test_size=0.25, random_state=12345)

features_train_r3, features_valid_r3, target_train_r3, target_valid_r3 = \
train_test_split(features_r3, target_r3, test_size=0.25, random_state=12345)


<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b>
    
<b>Успех:</b><br>
👍 Данные разделены на обучающую и валидационную выборку корректно - 75:25.<br>
Будет надежнее, если после разбиения проверять доли)    
</div>

In [59]:
scaler = StandardScaler() # сразу проскейлим выборку
scaler.fit(features_train_r1)
scaler.fit(features_train_r2)
scaler.fit(features_train_r3)

StandardScaler()

<div class="alert alert-block alert-warning"><b>Комментарий ревьюера</b>
    
<b>Совет:</b><br>
Павел, если бы ты стал применять `transform()` к выборкам, была бы ошибка. Получается `scaler` у тебя обучен на третьем регионе, а  `transform()` применил бы ко всем регионам, это некорректно.<br>
    
Правильно было бы так:<br>
```python    
scaler1 = StandardScaler() 
scaler1.fit(features_train_r1)
scaler2 = StandardScaler()    
scaler2.fit(features_train_r2)
scaler3 = StandardScaler()    
scaler3.fit(features_train_r3)
```
    
А затем применить  `scaler1, scaler2, scaler3` каждый к своему региону.<br>
А еще лучше создать функцию. <br>   
    
И еще совет 👇   <br> 
Для `трейна` можно использовать сразу `fit_transform()`. Она объединяет функции: `fit()` и `transform()`.
</div>

In [60]:
#Обучим модели:
model_r1 = LinearRegression()
model_r1.fit(features_train_r1, target_train_r1)
predicted_valid_r1 = model_r1.predict(features_valid_r1)
rmse_r1 = mean_squared_error(target_valid_r1, predicted_valid_r1) ** 0.5
print('Средний запас предсказываемого сырья:',predicted_valid_r1.mean())
print('RMSE модели 1 региона:',rmse_r1)

model_r2 = LinearRegression()
model_r2.fit(features_train_r2, target_train_r2)
predicted_valid_r2 = model_r2.predict(features_valid_r2)
rmse_r2 = mean_squared_error(target_valid_r2, predicted_valid_r2) ** 0.5
print('Средний запас предсказываемого сырья:',predicted_valid_r2.mean())
print('RMSE модели 2 региона:',rmse_r2)

model_r3 = LinearRegression()
model_r3.fit(features_train_r3, target_train_r3)
predicted_valid_r3 = model_r3.predict(features_valid_r3)
rmse_r3 = mean_squared_error(target_valid_r3, predicted_valid_r3) ** 0.5
print('Средний запас предсказываемого сырья:',predicted_valid_r3.mean())
print('RMSE модели 3 региона:',rmse_r3)

Средний запас предсказываемого сырья: 92.78915638280621
RMSE модели 1 региона: 37.853527328872964
Средний запас предсказываемого сырья: 69.17831957030432
RMSE модели 2 региона: 0.8920592647717013
Средний запас предсказываемого сырья: 94.86572480562035
RMSE модели 3 региона: 40.07585073246016


**Вывод:**
<div class="alert alert-block alert-info">
Средний запас сырья больше в 1 и 3 регионе, однако средняя квадратичная ошибка минимальна во втором регионе. А значит что модель во втором регионе работает лучше.

<div class="alert alert-block alert-success"><b>Комментарий ревьюера V-2</b>
    
<b>Успех:</b><br>
    👍 Молодец, что обращаешь внимание на размер ошибки, в двух других регионах она катастрофически большая - почти 40% от  среднего запаса. Если посмотреть корреляцию между признаками и целевой переменной, можно понять почему в одном из регионов ошибка минимальная:)<br>
Т.к. средний запас в каждом регионе ниже объема безубыточности, то качество предсказания имеет огромное значение - важно определить скважины с максимальным запасом.   
    
</div>  

<div class="alert alert-block alert-warning"><b>Комментарий ревьюера</b>
    
<b>Совет:</b><br>
Для каждого региона действия повторяются, в таких случаях удобно использовать функцию. Это позволит не писать один и тот же код несколько раз и избегать ошибок. В случае необходимых изменений, достаточно поправить функцию, а не вносить правки несколько раз. 
</div>

<div class="alert alert-block alert-danger"><b>Комментарий ревьюера</b>
    
<s> <b>Ошибка:</b><br>
Не забывай, пожалуйста, формулировать выводы к каждому блоку. Здесь важно обратить внимание на размер ошибки, в каком регионе она минимальная и о чем это говорит:)</s>
</div>

In [61]:
#Правильные ответы и предсказания:
true_valid_r1 = target_valid_r1.reset_index(drop=True)
predicted_valid_r1 = pd.Series(predicted_valid_r1)

true_valid_r2 = target_valid_r2.reset_index(drop=True)
predicted_valid_r2 = pd.Series(predicted_valid_r2)

true_valid_r3 = target_valid_r3.reset_index(drop=True)
predicted_valid_r3 = pd.Series(predicted_valid_r3)

<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b>
    
<b>Успех:</b><br>
👍 Очень хорошо, что не забываешь про индексы, теперь они соответствуют.  
</div>

## Подготовка к расчёту прибыли

Подготовим метрики из условия для расчета прибыли.

In [62]:
BUDGET = 10_000_000_000 # бюджет на разработку в регионе
BARREL = 450_000 # 1 единица сырья приносит 450_000 рублей дохода
points = 500 # исследуемых точек в регионе
best_points = 200 # отбираем 200 лучших точек
budget_for_one = BUDGET / best_points

<div class="alert alert-block alert-warning"><b>Комментарий ревьюера</b>
    
<b>Совет:</b><br>
* Согласно руководству по написанию кода на Python [PEP-8](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html#section-29) константы принято объявлять в начале проекта после импортов, затем загрузка данных. Так проще будет их найти и изменить, если возникнет такая необходимость.
* Для записи больших чисел и более удобного чтения, чтобы не запутаться в нулях, можно использовать еще такую запись:  `10e9`, что означает 10 * 10**9.    
</div>

In [63]:
# Посчитаем минимальное кол-во барралей при условии работы на 200 точках.
min_barrel_for_one = budget_for_one / BARREL
print(f'Минимальное кол-во для разработки 1 скважины - {min_barrel_for_one} тыс.бар.')

Минимальное кол-во для разработки 1 скважины - 111.11111111111111 тыс.бар.


<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b>
    
<b>Успех:</b><br>
👍 Достаточный объём сырья для безубыточной разработки новой скважины рассчитан верно.
</div>

In [64]:
# Сравним со средними значениями в регионах
region1['product'].mean(), region2['product'].mean(), region3['product'].mean()

(92.49968421774328, 68.82391591806625, 94.99834211933415)

Как мы видим - при работе на  200 точках (пока не лучших) и распределении средств между ними нам достаточно 111.11 тыс баррелей в каждой точке в среднем для безубыточной работе в регионе. Средние же показатели ниже, особенно во втором регионе и поэтому нам нужно выбрать самые прибыльные точки

<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b><br>
    
Верно. В каждом регионе средний объем сырья меньше достаточного объёма для безубыточности, поэтому точность прогноза имеет большое значение. <br>   

Есть 25 000 точек, из которых в результате надо выбрать 200. В результате,  среди этих скважин могут быть скважины с достаточным и недостаточным запасом сырья. <br>
    
Имеено техника "Bootstrap" позволяет оценить риск убытков при случайным отборе точек. 1000 раз отбирая случайным образом точки, считаем какой % случаев принесет убытки. 
    
   
</div>

## Расчёт прибыли и рисков 

In [65]:
def revenue(target, predictions):
    predictions_best = predictions.sort_values(ascending=False)
    profit = target[predictions_best.index][:best_points].sum() * BARREL
    income = profit - BUDGET
    return income

<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b>
    
<b>Успех:</b><br>
👍 Функция в порядке.
</div>

In [66]:
revenue(true_valid_r1, predicted_valid_r1)

3365187237.700287

Напишем функцию для техники будстреп

In [67]:
def butstrap(target, predictions):
    state = np.random.RandomState(12345)    
    profit = []
    for i in range(1000):
        target_subsample = target.sample(n=points, replace=True, random_state=state)
        probs_subsample = predictions[target_subsample.index]
        profit.append(revenue(target_subsample, probs_subsample))
    profit = pd.Series(profit)
    lower = int(profit.quantile(q=0.025)) # нижнее значение доверительного интервала
    upper = int(profit.quantile(q=0.975)) # верхнее значение доверительного интервала
    risk = len(profit[profit < 0] / len(profit) * 100) # риск убытков
    mean = profit.mean()
    print("Средняя выручка:", mean)
    print("нижнее значение доверительного интервала", lower)
    print("верхнее значение доверительного интервала", upper)
    print("риск убытков", risk,'%')

<div class="alert alert-block alert-success"><b>Комментарий ревьюера</b>
    
<b>Успех:</b><br>
    👍 Отлично, что при семплировании данных устанавливаешь `replace = True`, это необходимо чтобы наблюдение в выборку попало несколько раз, что позволяет добиться независимости при каждом выборе, иначе вероятность достать элемент из выборки будет с каждым разом увеличиваться. Чем больше размер подвыборки, тем выраженнее будет зависимость, поэтому без возврата не удастся получить разнообразные варианты, а подвыборку бОльшего размера вообще не получить.

Почему именно так. Вспомним, что выборка - это некоторое представление о генеральной совокупности, ее часть. Значит, что где-то в регионе существуют и другие потенциальные точки добычи нефти, о которых мы можем не знать, но они являются частью этой генеральной совокупности. В этом случае повторяющиеся наблюдения можно считать разными точками добычи из всей генеральной совокупности с близкими характеристиками.
    
</div>  

<div class="alert alert-block alert-warning"><b>Комментарий ревьюера</b>
    
<b>Совет:</b><br>
Риск лучше не округлять до целых, т.к. он меньше единицы
</div>

Посчитаем метрики для каждого региона

In [68]:
butstrap(true_valid_r1, predicted_valid_r1)

Средняя выручка: 409428038.6214359
нижнее значение доверительного интервала -131536028
верхнее значение доверительного интервала 944395582
риск убытков 71 %


In [69]:
butstrap(true_valid_r2, predicted_valid_r2)

Средняя выручка: 536902587.4874535
нижнее значение доверительного интервала 114527902
верхнее значение доверительного интервала 970787910
риск убытков 9 %


<div class="alert alert-block alert-success"><b>Комментарий ревьюера </b><br>
    

Что такое  `quantile()` очень хорошо объясняет пример - 0.5-й квантиль, это, по сути, медиана, т.е. функция `quantile(0.5)` должна вернуть такое значение, которое меньше ровно половины элементов массива.<br>  
`quantile(0.025)` находит такое значение прибыли,  меньше которого только 2,5% элементов. И если это значение прибыли положительное, значит и риск меньше 2,5%. <br>
Выбирая наугад 1000 раз 500 случайных скважин и отбирая из них 200 лучших, убыток получаем менее, чем в 2,5% случаев.  
    
   
</div>

In [70]:
butstrap(true_valid_r3, predicted_valid_r3)

Средняя выручка: 353790362.7309422
нижнее значение доверительного интервала -159194031
верхнее значение доверительного интервала 875652502
риск убытков 102 %


**Общий вывод**
Во втором регионе нет рисков и максимальная выручка. Доверительный интервал находится в положительном диапазоне.
Это самый лучший регион для добычи.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    
<b>Успех:</b><br>
    👍 Все расчеты правильные и регион выбран абсолютно верно.
    
</div>  

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>Общий вывод по проекту</u></b>
<font color='green'><br> 


Павел, приложено много усилий, ты - молодец! Крепкий проект! Есть несколько важных моментов, которые я отметила и которые можно было бы улучшить, подсказки ты найдешь в моих комментариях. 

**Что мне особенно понравилось в твоей работе:**:
 
* Есть знакомство с данными.
* Рассчитан средний запас предсказанного сырья и размер ошибки для каждого региона.
* Рассчитан достаточный объём сырья для безубыточной разработки новой скважины.    
* Написана функция для расчёта прибыли.    
* Рассчитаны риски, прибыль и доверительный интервал для каждого региона.     
* Код логичный и понятный. 
* Проект полностью соответствует плану!<br>


**На что стоит обратить внимание**:
    
* Осталось совсем немного поправить, а именно: добавить недостающий вывод. Подсказки ты найдешь в моих комментариях.




**Доработка займет у тебя совсем немного времени) Желаю удачи и жду твой проект на повторное ревью!**😉 

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован